### PR TITLE
nested modals using react-overlays

### PR DESCRIPTION
**Pull request by broncha**
Please check the functionality, if thats good, Ill cleanup.

### CODE REVIEW

File: src/ModalPopupComponent.coffee [hunk 1]
```diff
@@ -2,56 +2,118 @@ React = require 'react'
 ReactDOM = require 'react-dom'
 H = React.DOM
 _ = require 'lodash'
+Modal = require 'react-overlays/lib/Modal'
+className = require "react-overlays/node_modules/classnames"
```
Comment (grassick) [inline src/ModalPopupComponent.coffee]: Reaching inside node_modules is a big no-no (and doesn't work with my npm). Add `classnames` to package.json and use it normally instead.

**Comment by grassick** [comment]
Ok, the overlapping seems to work great!

Could you completely remove state+trigger from the control though? It should have an onClose event (prop) that fires when close is requested, and should be completely stateless. If it is rendered, it should be visible, always. 

The trigger was handy for the demo, but the final class will be much simpler and easier to use without it.

**Comment by broncha** [comment]
Mm thats what i was confused about. Aren't modals supposed to have closed
and opened state? Once you close the modal how will you open it back again!
We can sure have an onClose prop callback called when close is requested.

May be I'm missing somethig on how we intend to use the modal component.
But modals are mostly rendered as hidden and show/hide just changes its
visibility.

**Comment by grassick** [comment]
In general, React stuff should be stateless. Even the textbox just has a value and an onChange event and doesn't store state internally.

For the modals, the component that contains it will hold the state as to whether to display the modal or not. That way, it can decide whether to close the modal when the user clicks outside of it, for example. Doing it this way makes a lot fewer assumptions about how it can be used.

Here's an example:

```coffeescript
class EditValueComponent extends React.Component
  contructor: ->
    super
    @state = {
      editingValue: null
    }

  handleStartEditing: =>
    @setState(editingValue: @props.value)

  handleFinishEditing: =>
    # Decide if we want to keep value/close modal
    # Maybe do some validation of @state.editingValue
    @props.onChange(@state.editingValue)
    @setState(editingValue: null)

  render: ->
    H.div null,
      if @state.editingValue
        React.createElement(Modal, 
          onClose: @handleFinishEditing,
          React.createElement(SomeOtherEditingComponent,
            value: @state.editingValue
            onChange: (val) => @setState(editingValue: val)
          )
        )

      H.button onClick: @handleStartEditing, "Click to Edit"
```
In this way, the state is really clear and is in one place. If editingValue is set, we're editing and keeping track of the value being edited. We have full control over whether to allow the modal to be dismissed and can do other checks when it's requested. 

Does that make sense? Moving state to upper components is one of the trickiest things to get about React but it makes everything so very much simpler.

**Comment by broncha** [comment]
Yes, got that. I was thinking of that as a wrapper component with the
trigger and the modal both inside that. Ill clean it up.

**Comment by broncha** [comment]
@grassick 
Ok, so here is the thing. The `Modal` component from the overlay library takes a prop `show`, which is what every modals do. IMO modals should have and control their own state. I set up a sample as you described:

```javascript
class ModalSample extends React.Component
  constructor: ->
    super
    @state = {
      editing: false
    }

  startEditing:=>
    @setState(editing: true)

  finishEditing:=>
    @setState(editing: false)

  handleModalClose: =>
    @finishEditing()

  render: ->
    H.div null,
      H.a onClick: @startEditing,
        "Edit me"
      if @state.editing
        R ModalPopupComponent, { header: "OUTER", size: "large", onClose: @handleModalClose}
```

The problem here, initial rendering of the modal is handled by this component, which is fine. This component is deciding whether the modal should be shown. Now when the modal is displayed, we cant close it, because this component is still editing. 

The modals takes the event, `escape button press` or `click` outside the dialog, i.e. the backdrop. When we go this way, for the modal to be dismissed, `ModalSample` `editing` state should be false, which cannot happen since the modal takes over the events.

So in my opinion, modals should always be rendered, and shown or hidden as we require. Once the modal is shown, it becomes the first responder. When a close is requested, it closes itself and propagates to the calling component.

I get the part `if @state.editing` , but this way this component will need to be re-rendered for the modal to disappear. I hope I have been clear on what I am trying to explain.

**Comment by grassick** [comment]
Hi Rajesh,

There are some good reasons to not keep the state in the modal, as it duplicates state in more complex examples. The Modal component in the react-overlay doesn't keep the displayed state; it's just a building block as a dumb component. I really want ours to be the same. I'll try to write a more detailed explanation tomorrow if I can.

>The modals takes the event, escape button press or click outside the dialog, i.e. the backdrop. When we go this way, for the modal to be dismissed, ModalSample editing state should be false, which cannot happen since the modal takes over the events.

The `onClose` event of the modal triggers the editing state to false, which does the trick, no? onClose is fired by the modal which sets editing to false and it's all done, no?

> I get the part if @state.editing , but this way this component will need to be re-rendered for the modal to disappear. I hope I have been clear on what I am trying to explain.

Re-rendering the entire tree in React is really fast and normal, so I'm afraid I don't quite understand the issue yet.

Thanks for talking this through and best wishes,

Clayton

**Comment by grassick** [comment]
Perhaps I misunderstood you... are you saying that a component based on the ModalComponent should hold the state or that ModalComponent itself should hold the state? If you mean the first one, then I agree that is a decent design for a modal. Having a show prop could be handy for this case exactly, though the component could just return null when it should not be shown.

**Comment by broncha** [comment]
I didn't get any emails about your updates here! Anyways, the problem I was having was due to a simple typo in coffeescript. It has been fixed.

**Pull request merged**